### PR TITLE
CXX-2179 Convert change streams spec tests to unified test format

### PR DESCRIPTION
**Pull request by eramongodb**
## Description

This PR resolves: CXX-1891, CXX-2179, CXX-2239, CXX-2296, CXX-2438, CXX-2440, CXX-2457, CXX-2462, CXX-2476, CXX-2478, CXX-2495, CXX-2497, CXX-2499, CXX-2514, CXX-2547, and CXX-2564.

This PR fast-forwards the change streams unified test files to https://github.com/mongodb/specifications/commit/e1a513fe027afb7706485a1e79c53a96acc2ec22 and the unified format tests to https://github.com/mongodb/specifications/commit/08a07f62fd9d857f72de21cad4dcde63144ce12b. This includes an update to CRUD unified test files as part of CXX-2296, an update to collection management unified test files as part of CXX-2499, and a drive-by improvement to `options::change_stream` to address CXX-1891.

## New Features

This PR adds C++ Driver support for the `fullDocumentBeforeChange` option (CXX-2438) and the `comment` option (CXX-2564, CXX-2457, CXX-2476) to change streams.

## Unified Format Tests

In an effort to simplify the sync state of the `valid-pass` tests, this PR syncs the unified format `valid-pass` tests _in their entirety_ up to https://github.com/mongodb/specifications/commit/08a07f62fd9d857f72de21cad4dcde63144ce12b (CXX-2495), allowing the maximum supported JSON schema version to be bumped from 1.5.0 to 1.7.0.

The `entity-client-cmap-events.json` and `assertNumberConnectionsCheckedOut.json` tests are skipped as they are blocked in C Driver CMAP work (see CDRIVER-3525, CXX-2166).

## Unified Test Runner Bug Fixes

`saveResultAsEntity` was incorrectly being treated as a required field by `createChangeStream` and `upload` operations. The `create_change_stream` and `upload` functions were updated accordingly.

`iterate_until_document_or_error` for change streams was not compliant with change streams unified spec tests due to its iterator increment behavior. The prior "prepare for next call to this function" pattern violated `getMore` event expectations by prematurely executing `getMore` _twice_ on first iteration when only one is expected. The test runner was refactored to correctly increment stream iterators _exactly once_ per `iterateUntilDocumentOrError` operation and _not_ increment them otherwise. This also includes support for the possibility of indefinitely blocking given `iter == stream.end()` on subsequent iterations per CXX-2441 (AFAIK no unified test actually requires this, but implemented nevertheless for completion/correctness).

The `list*` operations were not correctly reporting the resulting array of documents. Additionally, `listCollections` did not support the `filter` option. The `list_collections`, `list_databases`, and `list_indexes` helper functions were added accordingly.

## New Unified Test Runner Features

The `ChangeStreamFatalError` error code name translation to error code `280` was added to support change streams unified tests. Support for the `comment`, `fullDocument`, `fullDocumentBeforeChange`, `clusteredIndex`, and `changeStreamPreAndPostImages` options were also added accordingly.

As part of the unified format test files sync, support for the `rename`, `modifyCollection`, `createFindCursor`, `iterateUntilDocumentOrError` (for cursors), `ignoreExtraEvents`, `ignoreResultAndError`, `fullDocument`, and `fullDocumentBeforeChange` operations were added, many of which (but not all) are required by change streams unified tests.

### CODE REVIEW

**Comment by eramongodb** [comment]
As a followup to CXX-2438, the collection management unified test files were also synced with https://github.com/mongodb/specifications/commit/d96789119036eaa56e4c015a1e914919e30cb379 as part of CXX-2499. This includes the addition of support for the `modifyCollection` operation, support for the `clusteredIndex` and `changeStreamPreAndPostImages` options, and additional unified test runner bugfixes.

**Review by rcsanchez97 — APPROVED** [review]

**Comment by eramongodb** [comment]
Added CXX-2547 to resolve Evergreen task failures when testing against latest server versions (6.0.1 and newer).

**Review by kevinAlbs — COMMENTED** [review]
This looks like a lot of progress on getting specification tests and the runner up to date.

**Review by kevinAlbs — APPROVED** [review]

File: src/mongocxx/test/CMakeLists.txt [hunk 1]
```diff
@@ -276,6 +263,8 @@ set_property(TEST unified_format_spec
   APPEND PROPERTY ENVIRONMENT "SESSION_UNIFIED_TESTS_PATH=${PROJECT_SOURCE_DIR}/../../data/sessions/unified/")
 set_property(TEST unified_format_spec
   APPEND PROPERTY ENVIRONMENT "CRUD_UNIFIED_TESTS_PATH=${PROJECT_SOURCE_DIR}/../../data/crud/unified")
+  set_property(TEST unified_format_spec
+    APPEND PROPERTY ENVIRONMENT "CHANGE_STREAMS_UNIFIED_TESTS_PATH=${PROJECT_SOURCE_DIR}/../../data/change-streams/unified")
 set_property(TEST unified_format_spec
   APPEND PROPERTY ENVIRONMENT "VERSIONED_API_TESTS_PATH=${PROJECT_SOURCE_DIR}/../../data/versioned-api")
 set_property(TEST unified_format_spec
```
Comment (vector-of-bool) [inline src/mongocxx/test/CMakeLists.txt:270]: Not important here, but these can all be flattened to a single `set_property(TEST ... APPEND PROPERTY ENVIRONMENT ...)` call, as you can list one or more tests at once, and any number of values to append.

**Review by vector-of-bool — APPROVED** [review]
Sorry to delay on this. LGTM!

One unimportant remark (about CMake, of course).

**Pull request merged**